[MRG] FIX sklearn.cluster.spectral_clustering fails with `np.matrix` input (#10993)

sinanh opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #10993 

#### What does this implement/fix? Explain your changes.
It fixes a small bug in `spectral_embedding` while finding the neighbors when the input array is a `np.matrix`. 

#### Any other comments?
 

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

sinanh committed:
fixes 10993

sinanh committed:
fixes 10993

sinanh committed:
Merge remote-tracking branch 'upstream/master' into fix_issue_10993

sinanh committed:
Update test_spectral.py

sinanh committed:
PEP8 fixes

sinanh committed:
PEP8 fixes

jnothman commented:
Not changing estimator_checks

jnothman commented:
Not changing estimator_checks?

sinanh commented:
At first I changed estimator_checks, then I noticed that was not necessary. Because after some testing, I noticed that all the clustering algorithms (e.g. KMeans, AgglomerativeClustering, MeanShift, DBSCAN etc.) work with np.matrix smoothly. Thus, I think the mentioned problem in this issue was specific to Spectral Clustering.

jnothman commented:
interesting. I would still be keen on a common test to ensure matrix works
across the board, for instance. Your test whether clusterers work was
unlikely to check that attributes stored are arrays and not matrices, so I
also have my doubts about how certain you are that they completely worked...

jnothman commented on sklearn/manifold/spectral_embedding_.py:
```diff
@@ -60,6 +60,8 @@ def _graph_connected_component(graph, node_id):
         for i in indices:
             if sparse.issparse(graph):
                 neighbors = graph[i].toarray().ravel()
+            elif isinstance(graph, np.matrix):
```
Can we just convert and store as graph at the top of the function?

jnothman reviewed (commented):
Otherwise this LGTM.

Though we should really have a generic test for consistency between class and function versions of clusterers that makes this check, instead of ad-hoc tests.

jnothman commented:
we would not want an expensive check in the function version. those should
be low overheard to enable reuse in other code.

sinanh committed:
ensure arrays and matrices are converted to a graph

sinanh committed:
ensure arrays and matrices are converted to a graph

amueller commented:
I thought we didn't support np.matrix, but I guess if ``check_array`` accepts it then it's fine?

rth commented on sklearn/manifold/spectral_embedding_.py:
```diff
@@ -44,6 +44,8 @@ def _graph_connected_component(graph, node_id):
         node
     """
     n_node = graph.shape[0]
+    graph = sparse.csc_matrix(graph)
```
This does not look right -- it is converted to CSR on the following line.

sinanh replied:
Thanks, I think you are right.

To be compliant with jnothman's suggestion on May 8th about converting and storing it as a graph at the top, I suggest the following actions:

Change line 48 to: `graph = sparse.csr_matrix(graph)`
Delete lines 49-51

amueller replied:
I'm a bit confused about this. @jnothman?

jnothman replied:
> To be compliant with jnothman's suggestion on May 8th about converting and storing it as a graph at the top, I suggest the following actions:

What I said was "Can we just convert and store as graph at the top of the function?"

What I meant was "Can we just convert the matrix to an array and assign it back to the `graph variable at the top of the function?"

sinanh replied:
> This does not look right -- it is converted to CSR on the following line.

Actually, since it must be a symmetric matrix, its row and column matrix representations are equivalent. Thus it was passing the tests.

sinanh committed:
Update spectral_embedding_.py

converting the graph to sparse.csc_matrix was passing the tests, because it must be a symmetric matrix. So its column matrix (csc) and row matrix (csr) are equivalent.

sinanh committed:
Update spectral_embedding_.py

converting the graph to sparse.csc_matrix was passing the tests, because it must be a symmetric matrix. So its column matrix (csc) and row matrix (csr) are equivalent.

sinanh committed:
for merge

sinanh committed:
fix

sinanh commented:
It was failing  the test due to a bug in Azure pipeline as fixed in #14689. I can create another pull request in sync with the fix if needed.

sinanh committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

cmarmo commented:
Hi @sinanh , thanks for your patience! Are you still interested in working on this PR? If so do you mind fixing conflicts? Thanks!

sinanh committed:
squash commits

sinanh committed:
remove column_transformer

sinanh commented:
> Hi @sinanh , thanks for your patience! Are you still interested in working on this PR? If so do you mind fixing conflicts? Thanks!

I fixed the conflicts .

cmarmo commented:
Thanks @sinanh! @jnothman , @glemaitre maybe you might want to have a look? Thanks!

jnothman commented on sklearn/manifold/_spectral_embedding.py:
```diff
@@ -44,6 +44,9 @@ def _graph_connected_component(graph, node_id):
         node.
     """
     n_node = graph.shape[0]
+    if isinstance(graph, np.matrix):
```
instead, can we just pass np.asarray(graph) to _graph_connected_component, instead of passing graph unchanged?

sinanh replied:
@jnothman OK, I did that. Also I noticed since **numpy 1.15**, it is advised not to use `np.matrix`. The following is written in the manual:

> "It is no longer recommended to use this class, even for linear algebra. Instead use regular arrays. The class may be removed in the future."

sinanh committed:
Merge remote-tracking branch 'upstream/master' into fix_issue_10993

sinanh committed:
pass as ndarray

cmarmo commented:
Hi @sinanh, I guess you closed as you are tired of waiting? I can't do more than relabeling as "waiting for review"... but I wonder if you could reconsider...thanks for your work, anyway.

sinanh commented:
> Hi @sinanh, I guess you closed as you are tired of waiting? I can't do more than relabeling as "waiting for review"... but I wonder if you could reconsider...thanks for your work, anyway.

@cmarmo thanks, the thing is from numpy 1.15 it is advised not to use np.matrix, so this issue is not an issue anymore I think. I guess both the issue and PR can be closed.

cmarmo commented:
@sinanh, understood: thanks for clarifying. If you are still interested in working on scikit-learn, feel free to pick one among the "[help wanted](https://github.com/scikit-learn/scikit-learn/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22)" issues. Thanks for your time.
